### PR TITLE
chore: Support substitute global variable in Spec level elements

### LINKED ISSUE #4894: `workflow.parameters` cannot be referenced in `synchronization` field
## Summary 

`workflow.parameters` cannot be used in `synchronization` (both mutex and semaphore) after upgrading from v2.11.7 to v2.12.4. I am using it in `WorkflowTemplate` if that matters.

## Diagnostics

What Kubernetes provider are you using? 

EKS v1.15.12-eks-31566f

What version of Argo Workflows are you running? 

v2.12.4

```yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: semaphore-ref
spec:
  entrypoint: whalesay
  arguments:
    parameters:
      - name: semaphoreConfigMapKeyRef
        value: ""
  synchronization:
    semaphore:
      configMapKeyRef:
        name: my-config
        key: "{{workflow.parameters.semaphoreConfigMapKeyRef}}"
  templates:
    - name: whalesay
      container:
        image: docker/whalesay:latest
        command: [cowsay]
        args: ["hello world"]
```

```
$ kubectl logs
Failed to acquire the synchronization lock. configmaps "{{workflow.parameters.semaphoreConfigMapKeyRef}}" not found
```

```yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: mutex
spec:
  entrypoint: whalesay
  arguments:
    parameters:
      - name: repository
        value: ""
  synchronization:
    mutex:
      name: "{{workflow.parameters.repository}}"
  templates:
    - name: whalesay
      container:
        image: docker/whalesay:latest
        command: [cowsay]
        args: ["hello world"]
```

```
Paste the logs from the workflow controller:
$ kubectl logs
time="2021-01-17T11:10:04.994Z" level=info msg="argo/Mutex/{{workflow.parameters.repository}} acquired by argo/picnic-argo-gh-m2ppc " mutex="argo/Mutex/{{workflow.parameters.repository}}"
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by sarabala1979**
Signed-off-by: Saravanan Balasubramanian <sarabala1979@gmail.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).
closes #4894

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -250,23 +247,6 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 			woc.markWorkflowFailed(ctx, msg)
 			return
 		}
-		validateOpts := validate.ValidateOpts{ContainerRuntimeExecutor: woc.getContainerRuntimeExecutor()}
-		wftmplGetter := templateresolution.WrapWorkflowTemplateInterface(woc.controller.wfclientset.ArgoprojV1alpha1().WorkflowTemplates(woc.wf.Namespace))
-		cwftmplGetter := templateresolution.WrapClusterWorkflowTemplateInterface(woc.controller.wfclientset.ArgoprojV1alpha1().ClusterWorkflowTemplates())
-
-		// Validate the execution wfSpec
```
Comment (sarabala1979) [inline workflow/controller/operator.go:257]: Moving to SetWorkflowSpec

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -3199,14 +3178,45 @@ func (woc *wfOperationCtx) setExecWorkflow() error {
...
+		}
+
+		// If we received conditions during validation (such as SpecWarnings), add them to the Workflow object
+		if len(*wfConditions) > 0 {
+			woc.wf.Status.Conditions.JoinConditions(wfConditions)
+			woc.updated = true
+		}
+	}
+	woc.setGlobalParameters(woc.execWf.Spec.Arguments)
+	err := woc.substituteGlobalVariables()
```
Comment (simster7) [inline workflow/controller/operator.go:3216]: Should these two be the same function?

Comment (sarabala1979) [reply workflow/controller/operator.go:3216]: No, One function is populating a variable in the globalParammap to substitute the value in all places like here and executeTemplate. 
The second one is a new function that will substitute the global variable values into WorkflowSpec.

**Review by simster7 — APPROVED** [review]

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -3206,14 +3185,45 @@ func (woc *wfOperationCtx) setExecWorkflow() error {
 		woc.execWf = &wfv1.Workflow{Spec: *woc.wf.Status.StoredWorkflowSpec.DeepCopy()}
 		woc.volumes = woc.execWf.Spec.DeepCopy().Volumes
 	} else if woc.controller.Config.WorkflowRestrictions.MustUseReference() {
-		return fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
+		err := fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
+		woc.log.WithError(err).Errorf("Unable to get Workflow Template Reference for workflow")
+		woc.markWorkflowError(ctx, err)
```
Comment (alexec) [inline workflow/controller/operator.go:3190]: Why isn’t return error enough? The caller should be marking the work flow as error

Comment (sarabala1979) [reply workflow/controller/operator.go:3190]: Some err will be `Error` and some are `Failure`. That's why it is making here. I will move the log into upper level

**Pull request merged**